src/serde_support.rs

```diff
@@ -0,0 +1,120 @@
+extern crate serde;
+extern crate serde_test;
+
+use super::*;
+
+use self::serde::{Deserialize, Deserializer};
+use self::serde::de::{SeqAccess, Visitor};
+use self::serde::export::fmt;
+use self::serde::export::PhantomData;
+use self::serde::ser::{Serialize, Serializer, SerializeSeq};
+
+impl<T> Serialize for CircularQueue<T> where T: Serialize {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where
+        S: Serializer {
+        let mut seq = serializer.serialize_seq(Some(self.capacity()))?;
+        for e in self.asc_iter() {
+            seq.serialize_element(e)?;
+        }
+        seq.end()
```

Comment:
Since `CircularQueue` has a fixed capacity which does not change during operation, it should serialize its capacity first, and read in the capacity during deserialization. It cannot be inferred from the sequence length, because:
- the queue can be half-full during serialization, in which case we still want to restore the correct capacity,
- for formats where the sequence length isn't known (e.g. JSON I guess) the best we can do is read in the elements into a Vec, then create a queue with capacity equal to the number of elements and transfer all of them over, which once again might not result in a queue with correct capacity.

Reply:
I believe this is what I am doing now.  On like 15, you see:

``
        let mut seq = serializer.serialize_seq(Some(self.capacity()))?;
``

which sets the length of the sequence as the capacity (not the actual length).  I could not find any information in the contract for serde as to how it handles the distinction, or if there is some fundamental difference between how the length and capacity are serialized.  However, the token stream includes an `end` token that indicates how long the sequence actually is.  So during de-serialization you see

``
       let mut values = CircularQueue::with_capacity(seq.size_hint().unwrap_or(0));
        while let Some(value) = seq.next_element()? {
            values.push(value);
        }
``


Also, if you take a look at the tests, you will see exactly the case you are asking about:

``
    fn serialization_works() {
        let mut q = CircularQueue::with_capacity(8);
        q.push(1);
        q.push(2);

        assert_tokens(&q, &[
            Token::Seq { len: Some(8) },
            Token::I32(1),
            Token::I32(2),
            Token::SeqEnd,
        ]);
    }
``

Here, the serialization token stream includes the capacity first (8), followed by only two items.  I think this is what you intended.  It seems the least surprising option to me.

Reply:
I believe this argument is there for serialization formats which must know the sequence length up-front (think binary format with a [length + length elements] format for arrays). It should actually be set to the real number of elements in the queue for the code to be correct. Formats like JSON ignore this number.

Reply:
Thanks for the explanation!  I missed that detail.  I now serialize to a struct with two fields, the first one is the capacity of the queue, and the second is a `vec` of values that are obtained from the `asc_iter`.  To deserialize, I wrote a custom deserializer, but I'm not entirely sure it can't be done by using the `serde` deserialize macro and deserializing to a struct representation of the serialized struct and then constructing from that.  In any case, I added tests for JSON and Bincode.